pkg/cvo/metrics.go

```diff
@@ -105,38 +115,84 @@ type asyncResult struct {
 	error error
 }
 
-// RunMetrics launches an server bound to listenAddress serving
-// Prometheus metrics at /metrics over HTTPS.  Continues serving
-// until runContext.Done() and then attempts a clean shutdown
-// limited by shutdownContext.Done().  Assumes runContext.Done()
-// occurs before or simultaneously with shutdownContext.Done().
-func RunMetrics(runContext context.Context, shutdownContext context.Context, listenAddress string, tlsConfig *tls.Config) error {
-	if tlsConfig == nil {
-		return errors.New("TLS configuration is required to serve metrics")
-	}
-
+func createHttpServer() *http.Server {
 	handler := http.NewServeMux()
 	handler.Handle("/metrics", promhttp.Handler())
 	server := &http.Server{
 		Handler: handler,
 	}
+	return server
+}
 
-	tcpListener, err := net.Listen("tcp", listenAddress)
+func shutdownHttpServer(parentCtx context.Context, svr *http.Server) {
+	ctx, cancel := context.WithTimeout(parentCtx, 5*time.Second)
+	defer cancel()
+	klog.Info("Shutting down metrics server so it can be recreated with updated TLS configuration.")
+	if err := svr.Shutdown(ctx); err != nil {
+		klog.Errorf("Failed to gracefully shut down metrics server during restart: %w", err)
+	}
+}
+
+func startListening(svr *http.Server, tlsConfig *tls.Config, lAddr string, resultChannel chan asyncResult) {
+	tcpListener, err := net.Listen("tcp", lAddr)
 	if err != nil {
-		return err
+		resultChannel <- asyncResult{name: "HTTPS server", error: err}
+	}
+	tlsListener := tls.NewListener(tcpListener, tlsConfig)
+	klog.Infof("Metrics port listening for HTTPS on %v", lAddr)
+	err = svr.Serve(tlsListener)
+	resultChannel <- asyncResult{name: "HTTPS server", error: err}
+}
+
+// RunMetrics launches a server bound to listenAddress serving
+// Prometheus metrics at /metrics over HTTPS.  Continues serving
+// until runContext.Done() and then attempts a clean shutdown
+// limited by shutdownContext.Done().  Assumes runContext.Done()
+// occurs before or simultaneously with shutdownContext.Done().
+// Also detects changes to metrics certificate files upon which
+// the metrics HTTP server is shutdown and recreated with a new
+// TLS configuration.
+func RunMetrics(runContext context.Context, shutdownContext context.Context, listenAddress, certFile, keyFile string) error {
+	var tlsConfig *tls.Config
+	if listenAddress != "" {
+		var err error
+		tlsConfig, err = makeTLSConfig(certFile, keyFile)
+		if err != nil {
+			return fmt.Errorf("Failed to create TLS config: %v", err)
```

Comment:
I realize this is just moved from existing code, but it's another place we can update from `%v` to `%w`.

```console
$ git show origin/pr/726 | grep '^+.*Errorf' | grep -v '%w'
+                       return fmt.Errorf("Failed to create TLS config: %v", err)
```

Reply:
Fixed in 0e6136c727b05.